algorithms/C/tree/binary-search-tree.c

```diff
@@ -0,0 +1,123 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+struct node{
+	struct node* left;
+	struct node* right;
+	int val;
+};
+
+struct node* createNode(int val){
+	struct node* n = (struct node*)malloc(sizeof(struct node));
+	n->val = val;
+	n->left = NULL;
+	n->right = NULL;
+	return n;
+}
+
+struct node* insert(struct node* root, int val){
+	if(root == NULL){
+		root = createNode(val);
+	}
+	else if(root->val < val){
+		root->right = insert(root->right, val);
+	}
+	else{
+		root->left = insert(root->left, val);
+	}
+
+	return root;
+}
+
+bool is_full(struct node* root){
+	if(root == NULL || (root->left == NULL && root->right == NULL)){
+		return true;
+	}
+	else if(root->left != NULL && root->right != NULL){
+		return is_full(root->left) && is_full(root->right);
+	}
+	return false;
+}
+
+int depth(struct node* root){
+	int d = 0;
+	while(root != NULL){
+		root = root->left;
+		d++;
+	}
+	return d;
+}
+
+bool is_perfect(struct node* root, int depth, int current_lvl){
+	if(root == NULL){
+		return true; 
+	}
+	else if(root->left == NULL && root->right == NULL){
+		return depth == current_lvl;
+	}
+	else if(root->left != NULL && root->right != NULL){
+		return is_perfect(root->left, depth, 1+current_lvl) &&
+		is_perfect(root->right, depth, 1+current_lvl);
+	}
+	
+	return false;
+}
+
+void print(struct node* root){
+	if(root != NULL){
+		print(root->left);
+		printf("%i ", root->val);
+		print(root->right);
+	}
+}
+
+void deallocate(struct node* root){
+	if(root != NULL){
+		deallocate(root->left);
+		deallocate(root->right);
+		free(root);
+	}
+}
+
+int main(){
+	struct node* root = NULL;
+
+	root = insert(root, 37);
+	root = insert(root, 19);
+	root = insert(root, 4);
+	root = insert(root, 22);
+	root = insert(root, 51);
+	root = insert(root, 55);
+	root = insert(root, 42);
+	
+	if(is_full(root)){
+		printf("BST is full\n");
+	}
+	else{
+		printf("BST is not full\n");
+	}
+	if(is_perfect(root, depth(root)-1, 0)){ 
+		printf("BST is Perfect\n");
+	}
+	else{
+		printf("BST is not Perfect\n");
+	}
+	print(root);
+	/*
+		Tree structure
+	       37
+	      /  \
+	     19  51
+	    / \  / \
+	   4  22 42 55
+
+		Output:
+			BST is full
+			BST is Perfect
+			4 19 22 37 42 51 55
+	*/
+	deallocate(root);
+	printf("\n%i", root->val);
```

Comment:
Even after i de-allocate the memory, i can access the value, why is tht?

Reply:
Read this [Accessing data after memory has been released](https://stackoverflow.com/questions/42588482/c-accessing-data-after-memory-has-been-freeed) this may help you...

Reply:
@Ashborn-SM that's because, deallocating memory just makes the memory block ( that your pointer is pointing to  ) available for program or OS to allocate it to someplace else, you pointer may still point to same memory location that it was pointing before, but accessing such memory location leads to [Undefined Behaviour](https://en.wikipedia.org/wiki/Undefined_behavior). Such pointers are famously known as [dangling pointers](https://en.wikipedia.org/wiki/Dangling_pointer)

Reply:
@Arsenic-ATG @UG-SEP thanks for the help, i understand how the memory is being re-used for some other purpose by the OS but when i try to set each tree node to NULL it is not giving me a segfault, why is that? shouldn't that be the case?

Reply:
this is a segment fault statement you may access the root because you are changing the root from deallocating function and in that function, you pass the value of root, not the address so the change you do in root there will not change it from the main function
Hope you understand

Reply:
> @Arsenic-ATG @UG-SEP thanks for the help, i understand how the memory is being re-used for some other purpose by the OS but when i try to set each tree node to NULL and try to access that it is not giving me a segfault, why is that? i mean shouldn't that be the case a segfault ?

not always 
A lot of things happening here are not defined by language standards, so it's either your OS or your language implementation which is deciding what should happen in such cases, in case of a segfault, it is mostly generated by your OS memory manager when a process tries to access an invalid or illegal memory address, and accessing a pointer after deallocating the memory it pointes to most likely will not generate such error since the pointer is still pointing to some real memory ( again it's not guaranteed and largely depends on how your implementation and OS manages memory ).

If you would have set the `root` node to `NULL` after deallocating it, then it would have most likely caused a segfault as generally NULL doesn't map to a valid memory location ( again exceptions do exist where it is a valid memory address like in [x86-real mode](https://wiki.osdev.org/Real_Mode) ).

Reply:
Thanks alot @Arsenic-ATG @UG-SEP. 

I am all done here, is there any changes i should i do?

Reply:
Thanks alot @Arsenic-ATG @UG-SEP.
I was passing the `root` by value ig thts why i couldn't set it to NULL.

I am all done here.

Reply:
@Ashborn-SM that great
Happy to help you 😊😊😊